cachi2/core/package_managers/bundler/parser.py

```diff
@@ -0,0 +1,157 @@
+import json
+import logging
+import re
+import subprocess
+from importlib.resources import files
+from typing import Optional, Union
+from urllib.parse import urlparse
+
+import pydantic
+
+from cachi2.core.errors import PackageManagerError, PackageRejected, UnexpectedFormat
+from cachi2.core.rooted_path import PathOutsideRoot, RootedPath
+from cachi2.core.utils import run_cmd
+
+log = logging.getLogger(__name__)
+
+GEMFILE = "Gemfile"
+GEMFILE_LOCK = "Gemfile.lock"
+
+
+class _GemMetadata(pydantic.BaseModel):
+    """
+    Base class for gem metadata.
+
+    Attributes:
+        name:       The name of the gem.
+        version:    The version of the gem.
+    """
+
+    name: str
+    version: str
+
+
+class GemDependency(_GemMetadata):
+    """
+    Represents a gem dependency.
+
+    Attributes:
+        source:     The source URL of the gem.
+        checksum:   The checksum of the gem.
+    """
+
+    source: str
+    checksum: Optional[str] = None
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this gem dependency."""
+        return NotImplemented
+
+
+class GitDependency(_GemMetadata):
+    """
+    Represents a git dependency.
+
+    Attributes:
+        url:        The URL of the git repository.
+        revision:   Commit hash.
+        branch:     The branch of the git repository.
+    """
+
+    url: str
+    revision: str
+    branch: Optional[str] = None
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this git dependency."""
+        return NotImplemented
+
+    @pydantic.field_validator("url")
+    def validate_git_url(cls, url: str) -> str:
+        """Validate that the git URL is a valid git URL."""
+        parsed_url = urlparse(url)
+        if parsed_url.scheme != "https":
```

Comment:
We should probably get a bit more defensive and raise an exception if `urlparse` fails to parse the URL even though bundler seems to fail correctly on an invalid URL:
```
/usr/share/gems/gems/bundler-2.2.33/lib/bundler/vendor/uri/lib/uri/rfc3986_parser.rb:67:in `split': bad Bundler::URI(is not Bundler::URI?): "https://gi#t%hub.com/3scale/json-schema.git" (Bundler::URI::InvalidURIError)
	from /usr/share/gems/gems/bundler-2.2.33/lib/bundler/vendor/uri/lib/uri/rfc3986_parser.rb:73:in `parse'
	from /usr/share/gems/gems/bundler-2.2.33/lib/bundler/vendor/uri/lib/uri/common.rb:234:in `parse'
	from /usr/share/gems/gems/bundler-2.2.33/lib/bundler/source/git.rb:305:in `uri_hash'
	from /usr/share/gems/gems/bundler-2.2.33/lib/bundler/source/git.rb:344:in `git_scope'
	from /usr/share/gems/gems/bundler-2.2.33/lib/bundler/source/git.rb:227:in `cache_path'
	from /usr/share/gems/gems/bundler-2.2.33/lib/bundler/source/git.rb:324:in `git_proxy'
	from /usr/share/gems/gems/bundler-2.2.33/lib/bundler/source/git.rb:236:in `revision'
	from /venv/lib64/python3.9/site-packages/cachi2/core/package_managers/bundler/scripts/lockfile_parser.rb:24:in `block in <main>'
	from /venv/lib64/python3.9/site-packages/cachi2/core/package_managers/bundler/scripts/lockfile_parser.rb:11:in `each'
	from /venv/lib64/python3.9/site-packages/cachi2/core/package_managers/bundler/scripts/lockfile_parser.rb:11:in `<main>'
```

Reply:
Yeah, apparently this is the case. I solved it by doing a `gem install json`. The problem was the amount of dependencies I had to install to get this done :smiling_face_with_tear::

```
# dnf install -y make gcc redhat-rpm-config
```

Reply:
Do you mean instead of `ValueError` ?

Reply:
Right, pretty much.